Move Carvel proposals to vmware-tanzu/carvel (#20)
vmware-tanzu/carvel: proposals/README.md

# Carvel Proposals
This directory serves as the home for Carvel proposals. A proposal is a design
document that describes a new feature for a Carvel project. The new feature
can span multiple projects or introduce a new project. Proposals are primarily
intended for maintainers but are open to the community for collaboration.

# Submit a Proposal
To create a proposal, submit a PR to this directory under the appropriate
project (for example, `ytt`). If the proposal concerns multiple projects or is
intended for the entire Carvel suite then please create the proposal at the root
of the `proposals` directory.

> 

Reviewer: ```suggestion

In that directory, create a `README.md` containing the core proposal. Include other files (e.g. sets of example artifacts, if necessary) to help support understanding of the feature.
```

**intention:** make plain that the proposal is a markdown file with a well-defined name for ease of discovery.

This directory serves as the home for Carvel proposals. A proposal is a design
document that describes a new feature for a Carvel project. The new feature
can span multiple projects or introduce a new project. 
A proposal must be sponsored by at least one Carvel Maintainer.  Proposals can be worked by anyone in the community.

# Submit a Proposal
To create a proposal, submit a PR to this directory under the appropriate
project with a terse name (for example, `ytt/001-schemas/`). If the proposal concerns multiple projects or is
intended for the entire Carvel suite then please create the proposal at the root
of the `proposals` directory (for example, `./010-carvel-cli/`).

In that directory, create a `README.md` containing the core proposal. Include other files (e.g. sets of example artifacts, if necessary) to help support understanding of the feature.
When creating the proposal, please add a `Status: Draft` line at the top of the
proposal to indicate its state.

## Proposal Template
The below template is an example. Other than the high-level details (such as
title, proposal status, author, and approvers), please use whichever sections
make the most sense for your proposal.

```md

> # <Proposal Title>
> - Status: *Draft* | In Review | Accepted | Rejected
> - Author(s): <your name>
> - Approvers: <list of names that must sign off>

Reviewer: ```suggestion
---
id: 1
title: "Writing a HIP"
authors: [ "Aaron Hurley <ahurley@vmware.com>" ]
status: "draft"
approvers: [ "Cari Dean <cdean@vmware.com", "Dmitriy Kalinin <dkalinin@vmware.com>" ]
---

# <Proposal Title>
```

**intention:** Use YAML metadata for metadata — it's recognized by tools like GitHub and hackmd (perhaps others?).

Author: Fancy. I ❤️ YAML.

> ## Problem Statement
> _This is a short summary of the problem that exists, why it needs to be
> solved, and how this proposal will address this need. The goal of this section is to bring
> readers up to speed quickly._

Reviewer: ```suggestion
## Problem Statement
_This is a short summary of the problem that exists, why it needs to be
solved: what specific needs are being met. Compelling problem statements
include concrete examples (even if only by reference). How exactly the proposal
would meet those needs should be located in the "Proposal" section, not this one.
The goal of this section is to help readers quickly empathize with the target users'
current experience to motivate the proposed change.
```

**intention:**
- encourage clarity by separating the problem from the solution (it's easy to blur the two).
- requesting concrete examples helps everyone involved tease apart "want" from "need".

Author: Good clarification and addition of the concrete examples 👍

## Terminology / Concepts
_Define any terms or concepts that are used throughout this proposal._

## Proposal
_This is the primary content of the proposal explaining how the problem(s) will
be addressed._

### Goals and Non-goals
_A short list of what the goals of this proposal are and are not._

### Specification / Use Cases
_Detailed explanation of the proposal's design._

> 

Reviewer: ```suggestion

### Other Approaches Considered
_Mention of other reasonable ways that the problem(s)
could be addressed with rationale for why they were less
desirable than the proposed approach._

```

**intention:**
- prompt for the negative solution space to ferret out any blind spots
- increase reviewers confidence that the solution space has been adequately considered
- rationale often illuminates the design thinking of the proposer which is a proxy for the quality of the proposal itself

Author: I like it!

# Carvel Proposals
This directory serves as the home for Carvel proposals. A proposal is a design
document that describes a new feature for a Carvel project. The new feature
can span multiple projects or introduce a new project. 

> A proposal must be sponsored by at least one Carvel Maintainer.  Proposals can be worked by anyone in the community.

Reviewer 2: When you say sponsored by a Carvel Maintainer do you mean that the proposal PR has to be created by a Maintainer, or that anyone can create a proposal after discussing it with a Carvel Maintainer?

Reviewer: > When you say sponsored by a Carvel Maintainer do you mean that the proposal PR has to be created by a Maintainer, or that anyone can create a proposal after discussing it with a Carvel Maintainer?

Generally speaking, I am thinking of trying to set boundaries as far out as possible to achieve that inflection point of: most inviting for participation with sufficient context.

So, yeah, I think anyone has the capacity to write a proposal. But "sponsored by a Carvel Maintainer" is actionable short-hand for, "it ought to be informed by the overall vision of the project."

Reviewer: @joaopapereira — does this sufficiently align with your thinking? If so, please resolve.

Reviewer 2: sounds good

# Submit a Proposal
To create a proposal, submit a PR to this directory under the appropriate
project with a terse name (for example, `ytt/001-schemas/`). If the proposal concerns multiple projects or is
intended for the entire Carvel suite then please create the proposal at the root
of the `proposals` directory (for example, `./010-carvel-cli/`).

In that directory, create a `README.md` containing the core proposal. Include other files (e.g. sets of example artifacts, if necessary) to help support understanding of the feature.
When creating the proposal, please add a `Status: Draft` line at the top of the
proposal to indicate its state.

## Proposal Template
The below template is an example. Other than the high-level details (such as
title, proposal status, author, and approvers), please use whichever sections
make the most sense for your proposal.

```md
---

> id: 1

Reviewer 2: Does this id need to match the name of the folder?
If so do we need to provide it if it is already in the folder name?

Reviewer: Great point: superfluous. Delete.

> title: "Writing a HIP"

Reviewer: ```suggestion
title: "Writing a Proposal"
```

Ooops! So, this history will help record our take-it-as-flattery theft from our friends at Helm... but it would be downright confusing to keep this in the template. 😬

authors: [ "Aaron Hurley <ahurley@vmware.com>" ]
status: "draft"

> approvers: [ "Cari Dean <cdean@vmware.com", "Dmitriy Kalinin <dkalinin@vmware.com>" ]

Reviewer 2: How might the PR creator know who the approvers should be?

Reviewer: the Carvel Maintainer sponsor should _at the very least_ know who might be minimally interested.

Default would prolly be the project approvers (publicly forth-coming Contribution Model should make that plain).

Reviewer: @joaopapereira — does this sufficiently align with your thinking? If so, please resolve.

Reviewer 2: sounds good

## Terminology / Concepts
_Define any terms or concepts that are used throughout this proposal._

## Proposal
_This is the primary content of the proposal explaining how the problem(s) will
be addressed._

### Goals and Non-goals
_A short list of what the goals of this proposal are and are not._

### Specification / Use Cases
_Detailed explanation of the proposal's design._

### Other Approaches Considered
_Mention of other reasonable ways that the problem(s)
could be addressed with rationale for why they were less
desirable than the proposed approach._

## Open Questions
_A list of questions that need to be answered._

## Answered Questions
_A list of questions that have been answered._
```

# Proposal Review
Once a proposal PR is submitted, project maintainers will review the proposal.
The goal of the review is to gain an understanding of the problem being solved
and the design of the proposed solution.

# Proposal States
| State | Definition |
| --- | --- |
| Draft | The proposal is actively being written by the proposer. |
| In Review | The proposal is being reviewed by project maintainers. |
| Accepted | The proposal has been accepted by the project maintainers. |

> | Rejected | The proposal has been rejected by the project maintainers. |

Reviewer 2: What happens if a proposal is rejected? Will the PR be merged with the "Rejected" state?

Reviewer: What we do _**not**_ do is sometimes just as important (or even more instructive) than what we actually do.

Also, someone proposing a similar idea some time later could reference prior art or recognize the impedance mismatch of the prior effort and adjust to make it more aligned with the project goals.

Reviewer: @joaopapereira — does this sufficiently align with your thinking? If so, please resolve.

Reviewer 2: Sounds good

> | State | Definition |
> | --- | --- |
> | Draft | The proposal is actively being written by the proposer. |
> | In Review | The proposal is being reviewed by project maintainers. |
> | Accepted | The proposal has been accepted by the project maintainers. |
> | Rejected | The proposal has been rejected by the project maintainers. |

Reviewer: ```suggestion
| Status | Definition |
| --- | --- |
| Draft | The proposal is actively being written by the proposer. |
| In Review | The proposal is being reviewed by project maintainers. |
| Accepted | The proposal has been accepted by the project maintainers. |
| Rejected | The proposal has been rejected by the project maintainers. |

## Lifecycle of a Proposal
1. Author adds a Proposal by:
   - creating a PR (in draft mode). (in this they can save their work until ready)
2. When the Author elaborates the proposal sufficiently to withstand critique they:
   - change the status to `in-review` and
   - mark the PR as "Ready for Review"
3. The community critiques the proposal (mature/converge it) by:
   - adding PR reviews.
4. When the Approvers reach [rough consensus](https://en.wikipedia.org/wiki/Rough_consensus), they:
   - change the status to `accepted` or `rejected`,
   - record both majority and dissenting opinions, and
   - merge the PR.
```

**intention:** make plain the conditions for status changes and the role of each actor in the lifecycle of the proposal.

Author: Nice addition.

document that describes a new feature for a Carvel project. The new feature
can span multiple projects or introduce a new project. 
A proposal must be sponsored by at least one Carvel maintainer.  Proposals can be worked by anyone in the community.

# Submit a Proposal
To create a proposal, submit a PR to this directory under the appropriate
project with a terse name (for example, `ytt/001-schemas/`). If the proposal concerns multiple projects or is
intended for the entire Carvel suite then please create the proposal at the root
of the `proposals` directory (for example, `./010-carvel-cli/`).

In that directory, create a `README.md` containing the core proposal. Include other files (e.g. sets of example artifacts, if necessary) to help support understanding of the feature.
When creating the proposal, please add a `Status: Draft` line at the top of the
proposal to indicate its state.

## Proposal Template
The below template is an example. Other than the high-level details (such as
title, proposal status, author, and approvers), please use whichever sections
make the most sense for your proposal.

```md
---

> id: 1

Reviewer 2: ```suggestion
```
